TINKERPOP-1752: Gremlin.Net: Generate completely type-safe methods

FlorianHockmann opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-1752

All steps are now (basically) type-safe and the original argument names from Gremlin-Java are used. However, we currently don't support some Java types like `Comparator`. Those were simply replaced by `object` until we find a better solution. A problem with this workaround is that some overloads from Gremlin-Java are not supported in Gremlin.Net as they would result in the same method signature.

The implementation of `Bindings` is now basically the same as in Gremlin-Java. This change was necessary as `Bindings.Of()` can no longer return a `Binding` object

These changes also revealed a bug in the tests where the `WithoutStrategies` source step was called with objects of strategies instead of just with their types. However, `WithoutStrategies` still can't work right now as a GraphSON serializer is missing for `Type`.

The `Group` step for example looks now like this:
```cs
/// <summary>
///     Adds the group step to this <see cref="GraphTraversal{SType, EType}" />.
/// </summary>
public GraphTraversal< S , IDictionary<K, V> > Group<K, V> ()
{
    Bytecode.AddStep("group");
    return Wrap< S , IDictionary<K, V> >(this);
}

/// <summary>
///     Adds the group step to this <see cref="GraphTraversal{SType, EType}" />.
/// </summary>
public GraphTraversal< S , E > Group (string sideEffectKey)
{
    Bytecode.AddStep("group", sideEffectKey);
    return Wrap< S , E >(this);
}
```
So this should also solve [TINKERPOP-1751](https://issues.apache.org/jira/browse/TINKERPOP-1751) that mentioned the missing overload for `Group` that returns `IDictionary<K, V>`.

spmallette commented:
@FlorianHockmann sorry but it looks like you have to rebase this one as I just merged #713 which moved the groovy scripts out of the pom.xml.  i didn't realize that change was going to affect this PR. I'll try to review as soon as you do that so that this PR doesn't go stale again. thanks.

FlorianHockmann commented:
@spmallette No problem, I could also have seen that by myself. To be honest, I just looked at the title of that PR and didn't recognize that it was about simplifying the GLV generation. In general, I really like the changes you introduced there as writing that much Groovy code inside of a pom.xml wasn't really nice.

FlorianHockmann commented:
Rebased on tp32, but now one of the two CI builds failed due to a failing test in `spark-gremlin`: `SparkContextStorageCheck.shouldSupportDirectoryFileDistinction`. From what I can see, the failing assert checks whether 10 files were actually created and here `ls` only returns 8 files.
I have honestly no idea how my changes could cause this test to fail as I didn't modify anything outside of `gremlin-dotnet`.

spmallette commented:
Travis is a flakey - I tend to think of a failure as a warning especially when one passes and one fails.  I wouldn't worry about it. If you did your own build and all is good then no problem.

FlorianHockmann commented:
Ok, good to know.
In addition to the Gremlin.Net tests, all tests pass with `docker/build.sh -t -i`.

jorgebay commented:
I wont be able to review it today, I'll probably have some time to look into this early next week.

spmallette commented:
Changes look generally ok to me, but I'll let @jorgebay do a more technical review of the code itself as i'm still trying to get my head back into .NET. Getting rid of the `params object[] args` seems like a helpful thing to do for users. Based on that: 

VOTE +1

jorgebay commented on gremlin-dotnet/glv/GraphTraversal.template:
```diff
@@ -65,9 +65,17 @@ namespace Gremlin.Net.Process.Traversal
         {
-            Bytecode.AddStep("<%= method.methodName %>", args);
+        <%  if (method.parameters.contains("params ")) {
+      %>    var args = new List<object> {<%= method.paramNames.init().join(", ") %>};
```
We should set the initial capacity of the list to avoid unnecessary resizing of the underlying array.
It should be the sum of the length of `method.paramNames.init()` and `method.paramNames.last()`.

jorgebay commented on gremlin-dotnet/src/Gremlin.Net/Process/Traversal/Bytecode.cs:
```diff
@@ -79,7 +85,77 @@ public void AddSource(string sourceName, params object[] args)
         {
-            StepInstructions.Add(new Instruction(stepName, args));
+            StepInstructions.Add(new Instruction(stepName, FlattenArguments(args)));
+            Bindings.Clear();
```
Even though `Bindings` is based on a `ThreadLocal<T>` instance, I think the implementation is not thread-safe.

For example: Multiple tasks executed serially on the same thread, modifying the same bindings dictionary.
Besides not being thread-safe, it doesn't support defining a binding on 1 thread and adding the step on another, sample:

```csharp
// thread 1
// define bindings
Bindings.Instance.Of(/* */);
await SomeUnRelatedTask();
// thread 2
// Do something with that binding
g.V()/**/;
```

FlorianHockmann replied:
This is definitely a problem, but do you see a way to fix it? Unfortunately, I think we have to use `static` here unless we want to create overloads for each parameter in the Traversal API that takes a `Binding` instead of the unbound variable which would result in a lot of overloads.

Since [`Bindings` don't seem to improve the performance with Bytecode](https://lists.apache.org/thread.html/c10baff72e67a8b7728c42b5c4e983a83e82b8f6964b3aa465f0e341@<dev.tinkerpop.apache.org>) when no lambdas are used which we don't support anyway, we may also remove the support for `Bindings` altogether from Gremlin.Net. Then we also don't need the `FlattenArguments` and `ConvertArgument` functions in `Bytecode` which are not exactly intuitive.

Or if we want to continue supporting them, we could also add a note to the documentation to make users aware of these concurrency problems with `Bindings`.

jorgebay replied:
I agree that a solution could be an overkill on the API...

I'm +1 from not supporting bindings in the .NET GLV (and document it), I don't see much purpose.

What do you think about not supporting Bindings on the .NET GLV @okram @spmallette @robertdale ?

spmallette replied:
Any reason we don't support lambdas? Even if .NET can't support them natively for some reason wouldn't we minimally support the ability to pass a python/groovy/etc lambda? it's kinda weird that way, but i think back to the point that kuppitz made on the dev list the other day where he stated that he doesn't always find a way out of using lambdas in production systems he works on - so ultimately users will need that kind of capability i think.

FlorianHockmann replied:
That's a good point. So we should probably support lambdas. To be honest, I haven't really looked into what would be necessary to support lambdas. If I'm not mistaken then we can easily support Python and Groovy lambdas when we add a dedicated GraphSON serializer for them and probably a class that wraps the lambda together with its language. Then the steps that take a lambda would simply accept an object of that class in Gremlin.Net. I can create a separate ticket for this when we agree that we want to support them.

When we want to support lambdas then we should also support `Bindings`. That leaves us with the question of whether we find a better implementation that doesn't suffer from the concurrency problems @jorgebay mentioned or if the current implementation in this pull request is acceptable.

Apart from that, shouldn't we clarify the section in the documentation about bindings? It currently states for example [for gremlin-python](http://tinkerpop.apache.org/docs/current/reference/#_bindings):

> If a traversal is going to be executed repeatedly, but with different parameters, then bindings should be used.

because:
> translation and compilation can be skipped as the previously compiled version should be cached

which is only true when a lambda is used if I understood it correctly.

jorgebay commented on gremlin-dotnet/src/Gremlin.Net/Process/Traversal/Bindings.cs:
```diff
@@ -29,14 +32,42 @@ namespace Gremlin.Net.Process.Traversal
+            var dict = BoundVariableByValue.Value;
+            if (dict == null)
+                return null;
+            return !dict.ContainsKey(value) ? null : dict[value];
```
Use `Dictionary{K, V}.TryGetValue()` to save one operation.

FlorianHockmann commented:
I finally found some time to work on this again and fixed the issues mentioned by @jorgebay. However, I left the `Bindings` implementation unchanged despite the problems with concurrent access as it seems to be still the best solution. (I'm of course open for suggestions on how this can be improved.)

I also noticed that my changes broke the `WithoutStrategies` source step as that now correctly expects to get the `Types` of the Strategies to exclude for which Gremlin.Net had no serializer. So I added a serializer that works basically like the respective one in gremlin-python as it also simply creates an object of the `Type` and then this object will be serialized as before. A unit test ensures that all Strategies have a parameterless constructor as we can't serialize their `Type` otherwise.
Honestly, I was a bit surprised that I had to serialize `Types` by serializing an empty object of that `Type` although the IO docs show that the GraphSON type `g:Class` [can be serialized like this](http://tinkerpop.apache.org/docs/3.3.0/dev/io/#_class):
```json
{
  "@type" : "g:Class",
  "@value" : "java.io.File"
}
```
but the Gremlin Server couldn't deserialize the Strategy class when I serialized it like this. So is the documentation wrong here? @spmallette: Could you clarify my confusion here?
Also the IO docs don't mention how `TraversalStrategies` are serialized in general. Should we add that?

The build is currently failing, but that seems to be caused by travis-ci/travis-ci#8607. I built Gremlin.Net locally and executed the tests without any problems.

BTW: Would it make sense to create a separate pull request for `master` or can we simply execute `generate.groovy` later when this is merged from `tp32` into `master`?

spmallette commented:
Looks like traversal strategy classes have their own serialization forms....................

```json
{
	"@type": "g:Bytecode",
	"@value": {
		"source": [
			["withStrategies", {
				"@type": "g:SubgraphStrategy",
				"@value": {
					"vertices": {
						"@type": "g:Bytecode",
						"@value": {
							"step": [
								["has", "name", "marko"]
							]
						}
					}
				}
			}]
		]
	}
}
```

I didn't know that. I guess the GraphSON spec in the IO docs (and testing framework) is incomplete. :rage1:

jorgebay commented:
VOTE +1

jorgebay commented:
All .NET tests passes with: 
```
mvn clean install -DskipTests
cd gremlin-dotnet && mvn clean install -P gremlin-dotnet
```

I'm running a `docker/build.sh -t -n -i`

spmallette commented:
VOTE +1

spmallette commented:
@jorgebay did the docker build ever finish successfully?

jorgebay commented:
Yes it builds, this PR is good to go on my end!

Sorry I forgot to confirm on Friday :/

jorgebay commented:
@FlorianHockmann Can we merge this?

FlorianHockmann commented:
> @FlorianHockmann Can we merge this?

Sure, I was on vacation, hence the delay. I'll merge it today.

FlorianHockmann commented:
I had to do a few changes during the merge into master which can be seen [in this commit](https://github.com/apache/tinkerpop/commit/ab2fd2cadd64a7af1438977ad9d31e9feff16eec), but now everything looks good and both branches build locally.
